### PR TITLE
feat(semantic): add  "unambiguous" `SourceType`

### PR DESCRIPTION
**Pull request by Boshen**
https://babel.dev/docs/options#sourcetype

### CODE REVIEW

**Comment by magic-akari** [comment]
I added a test case from Babel 😈

File: tasks/coverage/parser_typescript.snap [hunk 1]
```diff
@@ -4889,14 +4780,13 @@ Expect to Parse: tasks/coverage/typescript/tests/cases/conformance/salsa/plainJS
...
+Expect to Parse: tasks/coverage/typescript/tests/cases/conformance/salsa/privateIdentifierExpando.ts
 
-  × Cannot use export statement outside a module
-   ╭─[typescript/tests/cases/conformance/salsa/typeFromPropertyAssignmentWithExport.ts:3:1]
- 2 │ 
- 3 │ export const Adapter = {};
-   · ──────
- 4 │ 
+  × Private identifier '#bar' is not allowed outside class bodies
+   ╭─[typescript/tests/cases/conformance/salsa/privateIdentifierExpando.ts:2:3]
```
Comment (Boshen) [inline tasks/coverage/parser_typescript.snap:4786]: False postivie.

**Comment by Boshen** [comment]
I'll try and fix the issues in a stack.

**Comment by Boshen** [comment]
Reworked this in the stack https://github.com/oxc-project/oxc/pull/5557

**Pull request closed without merging**